### DISCUSSION TITLE
Cannot continue training model with forward attn batch norm

### DISCUSSION THREAD

**Discussion by JRMeyer** [coqui-ai/TTS#147] [Archived]
<USER>>>> georroussos<br/> <DATE> [February 22, 2020, 12:06pm]<br/> <br/> Cheers everyone. With the release of the model trained on Forward<br/> Attention and Batch Normalization afterwards, I thought I would see how<br/> it would fare when retraining on a new voice. I have a clean dataset,<br/> LJSpeech-ed (directories and metadata), however it refuses to train, on<br/> an assertion error:<br/> <br/> 12.53.58.filename}[1620×976 515<br/> <br/> I tried with LJSpeech instead, and got the same error. I am on the dev<br/> branch, that the model was also trained on and continue training using<br/> the model config and the checkpoint. I have decreased min_seq_length and<br/> have also tried removing any wavs shorter than 1 second, but it did not<br/> help. Would anyone have any tips? Thanks!<br/> <br/> [This is an archived TTS discussion thread from discourse.mozilla.org/t/cannot-continue-training-model-with-forward-attn-batch-norm]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> sanjaesc<br/> <DATE> [February 22, 2020, 12:41pm]<br/> <br/> Did you also remove the files from the metadata.csv?<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> georroussos<br/> <DATE> [February 22, 2020, 12:47pm]<br/> <br/> I did, yeah. I also tried with a small subset of 10 sentences that do<br/> not include any numbers and are longer than 1 second and it is still the<br/> same.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> nmstoker<br/> <DATE> [February 23, 2020, 12:48pm]<br/> <br/> Sorry for the obvious question, but what size is sentence35.wav in that<br/> exact directory path? (best to copy paste the path to be 100% sure) It<br/> does seem like the problem might be something else given your suggestion<br/> that you tried it with LJSpeech as well but that's the first step. Then<br/> look at the code in TTSDataset.py and work back from line 119<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> georroussos<br/> <DATE> [February 23, 2020, 1:10pm]<br/> <br/> No need to be sorry; the wav is 5 words long and some kb's. The problem<br/> isn't the file itself; the whole dataset isn't read (mine or LJSpeech).<br/> First, it complains about sentences where numbers are present. Then, I<br/> remove these sentences with a script (csv file and wav's). Then it<br/> complains about sentences with 1 word. Remove these and it complains<br/> about sentences with 2 words. I removed all sentences with length less<br/> than 5 and it still complains and I have also set the minimum seq length<br/> to 0. The files are not read. The model was trained on the dev branch,<br/> which is also the one I checked out in. This is my config:<br/> <br/> {<br/> <br/> 'github_branch': *slash* ' *slash* curl-run-all.sh discourse.mozilla.org html-to-markdown.sh ordered-posts ordered-posts~ TTS.cdx tts.commands tts-emails.txt TTS.pages tts-telegram.txt TTS.warc.gz dev *slash* ', *slash* <br/> //'restore_path': *slash* '/data/rw/pit/keep/ljspeech-December-11-2019_04+32PM-ca49ae8/checkpoint_410000.pth.tar *slash* ', *slash* <br/> 'github_branch': *slash* ' *slash* curl-run-all.sh discourse.mozilla.org html-to-markdown.sh ordered-posts ordered-posts~ TTS.cdx tts.commands tts-emails.txt TTS.pages tts-telegram.txt TTS.warc.gz dev *slash* ', *slash* <br/> 'model': 'Tacotron2', // one of the model in models/ *slash* <br/> 'run_name': 'ljspeech-bn', *slash* <br/> 'run_description': 'tacotron2 basline finetuned with BN prenet',<br/> <br/> // AUDIO PARAMETERS<br/> 'audio':{<br/> // Audio processing parameters<br/> 'trim_db': 0,<br/> 'mulaw': true,<br/> 'num_mels': 80, // size of the mel spec frame.<br/> 'num_freq': 1025, // number of stft frequency levels. Size of the linear spectogram frame.<br/> 'sample_rate': 22050, // DATASET-RELATED: wav sample-rate. If different than the original data, it is resampled.<br/> 'frame_length_ms': 10.0, // stft window length in ms.<br/> 'frame_shift_ms': 2.0, // stft window hop-lengh in ms.<br/> 'preemphasis': 0.98, // pre-emphasis to reduce spec noise and make it more structured. If 0.0, no -pre-emphasis.<br/> 'min_level_db': -100, // normalization range<br/> 'ref_level_db': 20, // reference level db, theoretically 20db is the sound of air.<br/> 'power': 1.5, // value to sharpen wav signals after GL algorithm.<br/> 'griffin_lim_iters': 60, // #griffin-lim iterations. 30-60 is a good range. Larger the value, slower the generation.<br/> // Normalization parameters<br/> 'signal_norm': true, // normalize the spec values in range [0, 1]<br/> 'symmetric_norm': true, // move normalization to range [-1, 1]<br/> 'max_norm': 4.0, // scale normalization to range [-max_norm, max_norm] or [0, max_norm]<br/> 'clip_norm': true, // clip normalized values into the range.<br/> 'mel_fmin': 50.0, // minimum freq level for mel-spec. ~50 for male and ~95 for female voices. Tune for dataset!!<br/> 'mel_fmax': 8000.0, // maximum freq level for mel-spec. Tune for dataset!!<br/> 'do_trim_silence': true // enable trimming of slience of audio as you load it. LJspeech (false), TWEB (false), Nancy (true)<br/> },<br/> <br/> // DISTRIBUTED TRAINING<br/> 'distributed':{<br/> 'backend': 'nccl',<br/> 'url': 'tcp: *slash* / *slash* /localhost:54321'<br/> },<br/> <br/> 'reinit_layers': [], // give a list of layer names to restore from the given checkpoint. If not defined, it reloads all heuristically matching layers.<br/> <br/> // TRAINING<br/> 'batch_size': 32, // Batch size for training. Lower values than 32 might cause hard to learn attention. It is overwritten by 'gradual_training'.<br/> 'eval_batch_size':16, <br/> 'r': 7, // Number of decoder frames to predict per iteration. Set the initial values if gradual training is enabled. <br/> 'gradual_training': [[0, 7, 64], [1, 5, 64], [50000, 3, 32], [130000, 2, 16], [290000, 1, 32]], // ONLY TACOTRON - set gradual training steps [first_step, r, batch_size]. If it is null, gradual training is disabled. <br/> 'loss_masking': true, // enable / disable loss masking against the sequence padding.<br/> <br/> // VALIDATION<br/> 'run_eval': true,<br/> 'test_delay_epochs': 10, //Until attention is aligned, testing only wastes computation time.<br/> 'test_sentences_file': null, // set a file to load sentences to be used for testing. If it is null then we use default english sentences.<br/> <br/> // OPTIMIZER<br/> 'noam_schedule': false,<br/> 'grad_clip': 1.0, // upper limit for gradients for clipping.<br/> 'epochs': 1000, // total number of epochs to train.<br/> 'lr': 0.0001, // Initial learning rate. If Noam decay is active, maximum learning rate.<br/> 'lr_decay': false, // if true, Noam learning rate decaying is applied through training.<br/> 'wd': 0.000001, // Weight decay weight.<br/> 'warmup_steps': 4000, // Noam decay steps to increase the learning rate from 0 to 'lr'<br/> 'seq_len_norm': false,<br/> <br/> // TACOTRON PRENET<br/> 'memory_size': -1, // ONLY TACOTRON - size of the memory queue used fro storing last decoder predictions for auto-regression. If < 0, memory queue is disabled and decoder only uses the last prediction frame.<br/> 'prenet_type': 'bn', // 'original' or 'bn'.<br/> 'prenet_dropout': false, // enable/disable dropout at prenet.<br/> <br/> // ATTENTION<br/> 'attention_type': 'original', // 'original' or 'graves'<br/> 'attention_heads': 5, // number of attention heads (only for 'graves')<br/> 'attention_norm': 'sigmoid', // softmax or sigmoid. Suggested to use softmax for Tacotron2 and sigmoid for Tacotron.<br/> 'windowing': false, // Enables attention windowing. Used only in eval mode.<br/> 'use_forward_attn': false, // if it uses forward attention. In general, it aligns faster.<br/> 'forward_attn_mask': false, // Additional masking forcing monotonicity only in eval mode.<br/> 'transition_agent': false, // enable/disable transition agent of forward attention.<br/> 'location_attn': true, // enable_disable location sensitive attention. It is enabled for TACOTRON by default.<br/> 'bidirectional_decoder': false, // use https://arxiv.org/abs/1907.09006. Use it, if attention does not work well with your dataset.<br/> <br/> // STOPNET<br/> 'stopnet': true, // Train stopnet predicting the end of synthesis.<br/> 'separate_stopnet': true, // Train stopnet seperately if 'stopnet==true'. It prevents stopnet loss to influence the rest of the model. It causes a better model, but it trains SLOWER.<br/> <br/> // TENSORBOARD and LOGGING<br/> 'print_step': 25, // Number of steps to log traning on console.<br/> 'save_step': 10000, // Number of training steps expected to save traninpg stats and checkpoints.<br/> 'checkpoint': true, // If true, it saves checkpoints per 'save_step'<br/> 'tb_model_param_stats': false, // true, plots param stats per layer on tensorboard. Might be memory consuming, but good for debugging.<br/> <br/> // DATA LOADING<br/> 'text_cleaner': 'phoneme_cleaners',<br/> 'enable_eos_bos_chars': false, // enable/disable beginning of sentence and end of sentence chars.<br/> 'num_loader_workers': 5, // number of training data loader processes. Don't set it too big. 4-8 are good values.<br/> 'num_val_loader_workers': 4, // number of evaluation data loader processes.<br/> 'batch_group_size': 0, //Number of batches to shuffle after bucketing.<br/> 'min_seq_len': 1, // DATASET-RELATED: minimum text length to use in training<br/> 'max_seq_len': 150, // DATASET-RELATED: maximum text length<br/> <br/> // PATHS<br/> 'output_path': '/data/rw/pit/keep/', // DATASET-RELATED: output path for all training outputs.<br/> //'output_path': '/media/erogol/data_ssd/Models/runs/',<br/> <br/> // PHONEMES<br/> 'phoneme_cache_path': 'ljspeech_ph_cache', // phoneme computation is slow, therefore, it caches results in the given folder.<br/> 'use_phonemes': true, // use phonemes instead of raw characters. It is suggested for better pronounciation.<br/> 'phoneme_language': 'en-us', // depending on your target language, pick one from https://github.com/bootphon/phonemizer#languages<br/> // MULTI-SPEAKER and GST<br/> 'use_speaker_embedding': false, // use speaker embedding to enable multi-speaker learning.<br/> 'style_wav_for_test': null, // path to style wav file to be used in TacotronGST inference.<br/> 'use_gst': false, // TACOTRON ONLY: use global style tokens<br/> <br/> // DATASETS<br/> 'datasets': // List of datasets. They all merged and they get different speaker_ids.<br/> [<br/> {<br/> 'name': 'ljspeech',<br/> 'path': '/home/georgios_roussos/TTS/LJSpeech-1.1/',<br/> //'path': '/home/erogol/Data/LJSpeech-1.1',<br/> 'meta_file_train': 'metadata_train.csv',<br/> 'meta_file_val': 'metadata_val.csv'<br/> }<br/> ]<br/> <br/> }<br/> <br/> I would really like to see how the model performs when re-trained on a<br/> new voice, because the performance it achieves on LJSpeech is very<br/> impressive and I am doing research on speaker adaptation.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> georroussos<br/> <DATE> [February 24, 2020, 12:09pm]<br/> <br/> Hi again, for anyone that might possibly struggle with it ﹣I found the<br/> config.json was the culprit; it would train okay with the initial config<br/> from the branch. The settings I changed on the checkpoint one were<br/> do_trim_db to 60 (I don't think it was there), added do_trim_silence and<br/> disabled prenet_dropout (I think). So, good to play with these in case<br/> problems occur.<br/> <br/> Shall return with updates on performance in regards to tuning!<br/> <br/> [Archived Post]<br/>
